builtin/v12/market/cbor_gen.go

```diff
@@ -1784,6 +1802,156 @@ func (t *OnMinerSectorsTerminateParams) UnmarshalCBOR(r io.Reader) (err error) {
 	return nil
 }
 
+var lengthBufSectorDealIDs = []byte{129}
+
+func (t *SectorDealIDs) MarshalCBOR(w io.Writer) error {
+	if t == nil {
+		_, err := w.Write(cbg.CborNull)
+		return err
+	}
+
+	cw := cbg.NewCborWriter(w)
+
+	if _, err := cw.Write(lengthBufSectorDealIDs); err != nil {
+		return err
+	}
+
+	// t.DealIDs ([]abi.DealID) (slice)
+	if len(t.DealIDs) > cbg.MaxLength {
+		return xerrors.Errorf("Slice value in field t.DealIDs was too long")
+	}
+
+	if err := cw.WriteMajorTypeHeader(cbg.MajArray, uint64(len(t.DealIDs))); err != nil {
+		return err
+	}
+	for _, v := range t.DealIDs {
+		if err := cw.CborWriteHeader(cbg.MajUnsignedInt, uint64(v)); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (t *SectorDealIDs) UnmarshalCBOR(r io.Reader) (err error) {
+	*t = SectorDealIDs{}
+
+	cr := cbg.NewCborReader(r)
+
+	maj, extra, err := cr.ReadHeader()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err == io.EOF {
+			err = io.ErrUnexpectedEOF
+		}
+	}()
+
+	if maj != cbg.MajArray {
+		return fmt.Errorf("cbor input should be of type array")
+	}
+
+	if extra != 1 {
+		return fmt.Errorf("cbor input had wrong number of fields")
+	}
+
+	// t.DealIDs ([]abi.DealID) (slice)
+
+	maj, extra, err = cr.ReadHeader()
+	if err != nil {
+		return err
+	}
+
+	if extra > cbg.MaxLength {
+		return fmt.Errorf("t.DealIDs: array too large (%d)", extra)
+	}
+
+	if maj != cbg.MajArray {
+		return fmt.Errorf("expected cbor array")
+	}
+
+	if extra > 0 {
+		t.DealIDs = make([]abi.DealID, extra)
+	}
+
+	for i := 0; i < int(extra); i++ {
+
+		maj, val, err := cr.ReadHeader()
+		if err != nil {
+			return xerrors.Errorf("failed to read uint64 for t.DealIDs slice: %w", err)
+		}
+
+		if maj != cbg.MajUnsignedInt {
+			return xerrors.Errorf("value read for array t.DealIDs was not a uint, instead got %d", maj)
+		}
+
+		t.DealIDs[i] = abi.DealID(val)
+	}
+
+	return nil
+}
+
+var lengthBufHamtCid = []byte{129}
+
+func (t *HamtCid) MarshalCBOR(w io.Writer) error {
+	if t == nil {
+		_, err := w.Write(cbg.CborNull)
+		return err
+	}
+
+	cw := cbg.NewCborWriter(w)
+
+	if _, err := cw.Write(lengthBufHamtCid); err != nil {
+		return err
+	}
+
+	// t.Cid (cid.Cid) (struct)
+
+	if err := cbg.WriteCid(cw, t.Cid); err != nil {
+		return xerrors.Errorf("failed to write cid field t.Cid: %w", err)
+	}
+
+	return nil
+}
+
+func (t *HamtCid) UnmarshalCBOR(r io.Reader) (err error) {
+	*t = HamtCid{}
+
+	cr := cbg.NewCborReader(r)
+
+	maj, extra, err := cr.ReadHeader()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err == io.EOF {
+			err = io.ErrUnexpectedEOF
+		}
+	}()
+
+	if maj != cbg.MajArray {
```

Comment:
This isn't quite the right serialization for a cid.  By default cbor gen is going to see a struct and serialize it as an array.  If you want a HamtCid type that will cborgen correctly by default you could make an alias type.  

While I sympathize with the instinct behind the special type given that our untyped raw pointers are not great I would probably avoid breaking our not-so-great convention.  I would avoid specially typing HamtCid altogether and just use the not-so-great convention of a plain cid.